cupyx/scipy/sparse/compressed.py

```diff
@@ -634,38 +687,111 @@ def __setitem__(self, slices, value, cs_type):
 
         major, minor = self._swap(row, col)
         major_size, minor_size = self._swap(*self._shape)
-        if numpy.isscalar(major):
-            i = int(major)
-            if i < 0:
-                i += major_size
-            if not (0 <= i < major_size):
-                raise IndexError('index out of bounds')
-            if numpy.isscalar(minor):
-                j = int(minor)
-                if j < 0:
-                    j += minor_size
-                if not (0 <= j < minor_size):
-                    raise IndexError('index out of bounds')
-                self._set_single(i, j, value, cs_type)
-        else:
-            raise ValueError('unsupported assignment')
-
-    def _set_single(self, major, minor, value, cs_type):
-        start = self.indptr[major]
-        end = self.indptr[major + 1]
-        data = self.data[start:end]
-        indices = self.indices[start:end]
-        if cs_type == 'csc':
-            index = len(data) - 1 - minor
-        elif cs_type == 'csr':
-            index = minor
-        else:
-            raise ValueError('unsupported class')
+        # flatten rows, cols, values lists
+        major = cupy.array(major, dtype=self.indices.dtype,
+                           copy=False, ndmin=1).ravel()
+        minor = cupy.array(minor, dtype=self.indices.dtype,
+                           copy=False, ndmin=1).ravel()
+        value = cupy.array(value, dtype=self.dtype,
+                           copy=False, ndmin=1).ravel()
+        # check boundaries
+        major_max = cupy.amax(major)
+        major_min = cupy.amin(major)
+        minor_max = cupy.amax(minor)
+        minor_min = cupy.amin(minor)
+        if major_max >= major_size or major_min < - major_size \
```

Comment:
This is causing unnecessary device synchronization which can greatly hurt performance.
Since you have the actual major and minor and shape objects in the cpu main memory, you can use them for comparisons.

Reply:
Actually major and minor here are arrays of rows and columns indices used in list assignment

Reply:
You can calculate them in the python objects,
doing if comparisons of cupy values force device synchronization and it can be pretty expensive.